docs/smart-contracts/meebits_erc721.sol

```diff
@@ -0,0 +1,679 @@
+
+    uint internal numTokens = 0;
+    uint internal numSales = 0;
+
+    // Cryptopunks contract
+    address internal punks;
+
+    // Autoglyphs contract
+    address internal glyphs;
+
+    address payable internal deployer;
+    address payable internal beneficiary;
+    bool public communityGrant = true;
+    bool public publicSale = false;
+    uint private price;
+    uint public saleStartTime;
+    uint public saleDuration;
+
+    //// Random index assignment
+    uint internal nonce = 0;
+    uint[TOKEN_LIMIT] internal indices;
+
+    //// Market
+    bool public marketPaused;
+    bool public contractSealed;
+    mapping (address => uint256) public ethBalance;
+    mapping (bytes32 => bool) public cancelledOffers;
+
+    modifier onlyDeployer() {
+        require(msg.sender == deployer, "Only deployer.");
+        _;
+    }
+
+    bool private reentrancyLock = false;
+
+    /* Prevent a contract function from being reentrant-called. */
+    modifier reentrancyGuard {
+        if (reentrancyLock) {
+            revert();
+        }
+        reentrancyLock = true;
+        _;
+        reentrancyLock = false;
+    }
+
+    modifier canOperate(uint256 _tokenId) {
+        address tokenOwner = idToOwner[_tokenId];
+        require(tokenOwner == msg.sender || ownerToOperators[tokenOwner][msg.sender], "Cannot operate.");
+        _;
+    }
+
+    modifier canTransfer(uint256 _tokenId) {
+        address tokenOwner = idToOwner[_tokenId];
+        require(
+            tokenOwner == msg.sender
+            || idToApproval[_tokenId] == msg.sender
+            || ownerToOperators[tokenOwner][msg.sender], "Cannot transfer."
+        );
+        _;
+    }
+
+    modifier validNFToken(uint256 _tokenId) {
+        require(idToOwner[_tokenId] != address(0), "Invalid token.");
+        _;
+    }
+
+    constructor(address _punks, address _glyphs, address payable _beneficiary) {
+        supportedInterfaces[0x01ffc9a7] = true; // ERC165
+        supportedInterfaces[0x80ac58cd] = true; // ERC721
+        supportedInterfaces[0x780e9d63] = true; // ERC721 Enumerable
+        supportedInterfaces[0x5b5e139f] = true; // ERC721 Metadata
+        deployer = msg.sender;
+        punks = _punks;
+        glyphs = _glyphs;
+        beneficiary = _beneficiary;
+    }
+
+    function startSale(uint _price, uint _saleDuration) external onlyDeployer {
+        require(!publicSale);
+        price = _price;
+        saleDuration = _saleDuration;
+        saleStartTime = block.timestamp;
+        publicSale = true;
+        emit SaleBegins();
+    }
+
+    function endCommunityGrant() external onlyDeployer {
+        require(communityGrant);
+        communityGrant = false;
+        emit CommunityGrantEnds();
+    }
+
+    function pauseMarket(bool _paused) external onlyDeployer {
+        require(!contractSealed, "Contract sealed.");
+        marketPaused = _paused;
+    }
+
+    function sealContract() external onlyDeployer {
+        contractSealed = true;
+    }
```

Comment:
Four functions only accessible for the deployer of the contract used to control the minting process and to provide security mechanisms